src/diffusers/models/attention_processor.py

```diff
@@ -761,12 +768,22 @@ class XFormersAttnProcessor:
     def __init__(self, attention_op: Optional[Callable] = None):
         self.attention_op = attention_op
 
-    def __call__(self, attn: Attention, hidden_states, encoder_hidden_states=None, attention_mask=None):
-        batch_size, sequence_length, _ = (
+    def __call__(
+        self,
+        attn: Attention,
+        hidden_states: torch.FloatTensor,
+        encoder_hidden_states: Optional[torch.FloatTensor] = None,
+        attention_mask: Optional[torch.FloatTensor] = None,
+    ):
+        batch_size, key_tokens, _ = (
             hidden_states.shape if encoder_hidden_states is None else encoder_hidden_states.shape
         )
 
-        attention_mask = attn.prepare_attention_mask(attention_mask, sequence_length, batch_size)
+        attention_mask = attn.prepare_attention_mask(attention_mask, key_tokens, batch_size)
+        if attention_mask is not None:
+            # xformers doesn't broadcast for us, so we expand our singleton dimension manually
+            _, query_tokens, _ = hidden_states.shape
+            attention_mask = attention_mask.expand(-1, query_tokens, -1)
```

Comment:
For my own understanding, is this ok because the mask is guaranteed to  (pre-expansion) be (batch size, 1, value sequence length). And then it's applied over the attention map of (batch size, key sequence length, value sequence length)?

Reply:
A comment here might be helpful, as I did have to think about this for a minute or so just to check :)

Reply:
this avoids an xformers error:

```
Expected bias.size(-2) == n_queries to be true, but got false.  (Could this error message be improved?  If so, please report an enhancement request to PyTorch.)
  File "/home/birch/git/pytorch/torch/_ops.py", line 646, in __call__
    return self._op(*args, **kwargs or {})
  File "/home/birch/git/diffusers-play/src/diffusers/src/diffusers/models/attention_processor.py", line 862, in __call__
    hidden_states = xformers.ops.memory_efficient_attention(
```

ordinarily, the user submits an attention mask of this shape to `UNet2dConditionModel#forward`:  
```
encoder_attention_mask.shape
[2, 77]
# [batch, tokens]
```

`UNet2dConditionModel#forward` adds a singleton dimension to help with the broadcasting we'll do later:  
```
encoder_attention_mask.shape
[2, 1, 77]
```

now `attention_mask` arrives in `AttnProcessor#forward`.  
`attn.prepare_attention_mask()` repeat_interleaves the batch dim per attention head:
```
attention_mask.shape
torch.Size([16, 1, 77])
# batch*heads, query_tokens, key tokens
```

ordinarily, this is all we need, to be able to add it as a bias to the following attention scores:  
```
torch.Size([16, 4096, 77])
```

for non-xformers implementations: that singleton query_tokens dimension is sufficient to broadcast this bias over each query token's attention score.

xformers does not implement broadcasting, so we have to manually expand the mask's singleton query_tokens dim over all the query tokens in the attention scores.

I'll see what I can add to the comment to help with that.

=====

extra detail:  
I've lied slightly about `tokens=77` in these examples. that was just for familiarity with CLIP dimensions.  
the reality is that as a user: I cannot submit a 77-token mask. xformers currently only supports token lengths with multiples of 8.  
as a user, this is what I currently do to make my CLIP condition and mask compatible with that constraint:  
https://github.com/Birch-san/diffusers-play/blob/700b31b758fe5148fc7c8c9083e01737027a11f7/scripts/play.py#L509-L524

Reply:
added comments explaining the mask's `query_tokens` dim, into `UNet2DConditionModel#forward` (which creates the singleton dim) and into `XFormersAttnProcessor#__call__` (the one place which is forced to expand the singleton dim).